### PR TITLE
Fix flaky tests in CI

### PR DESCRIPTION
**Pull request by crusaderky**
Attempt to fix some of the recent CI flakiness

### CODE REVIEW

File: distributed/diagnostics/tests/test_graph_layout.py [hunk 1]
```diff
@@ -38,16 +38,13 @@ async def test_construct_after_call(c, s, a, b):
...
-
-    updates = {state for idx, state in gl.state_updates}
-    assert "memory" in updates
-    assert "processing" in updates
-    assert "released" in updates
+    while True:
+        updates = {state for _, state in gl.state_updates}
+        if updates == {"waiting", "processing", "memory", "released"}:
+            break
+        await asyncio.sleep(0.01)
```
Comment (crusaderky) [inline distributed/diagnostics/tests/test_graph_layout.py:47]: test was randomly failing as gl.state_updates was tested before the "released" status appeared

File: distributed/tests/test_client_executor.py [hunk 2]
```diff
@@ -116,7 +107,7 @@ def test_cancellation(client):
 
 def test_cancellation_wait(client):
     with client.get_executor(pure=False) as e:
-        fs = [e.submit(slowinc, i, delay=0.1) for i in range(10)]
+        fs = [e.submit(slowinc, i, delay=0.2) for i in range(10)]
```
Comment (crusaderky) [inline distributed/tests/test_client_executor.py:110]: test was randomly failing as the cancellation occasionally took more than 0.1s to be enacted and thus one of the tasks had enough time to finish

File: distributed/tests/test_client_executor.py [hunk 3]
```diff
@@ -225,32 +216,30 @@ def test_retries(client):
...
-            with pytest.raises(RuntimeError):
-                e.submit(time.sleep, 1.0)
-
-            # shutdown(wait=False) cancels pending tasks
-            e = client.get_executor()
-            fut = e.submit(time.sleep, 2.0)
-            t1 = time.time()
-            e.shutdown(wait=False)
-            dt = time.time() - t1
-            assert dt < 0.5
```
Comment (crusaderky) [inline distributed/tests/test_client_executor.py:251]: test was randomly failing here as the shutdown took ~0.51s

File: distributed/tests/test_diskutils.py [hunk 4]
```diff
@@ -275,12 +275,14 @@ def _test_workspace_concurrency(tmpdir, timeout, max_procs):
...
 @pytest.mark.slow
 @pytest.mark.xfail(condition=MACOS, reason="extremely flaky")
+@pytest.mark.flaky(condition=not MACOS, reruns=10, reruns_delay=5)
 def test_workspace_concurrency(tmpdir):
     _test_workspace_concurrency(tmpdir, 5.0, 6)
 
 
 @pytest.mark.slow
 @pytest.mark.xfail(condition=MACOS, reason="extremely flaky")
+@pytest.mark.flaky(condition=not MACOS, reruns=10, reruns_delay=5)
```
Comment (crusaderky) [inline distributed/tests/test_diskutils.py:285]: recently found to have become highly unstable on Linux too

Comment (fjetter) [reply distributed/tests/test_diskutils.py:285]: Locally these tests take about 5 and 8 seconds for me. On CI this is likely not as fast and rerunning them for 10 times might increase runtime of our test suite by multiple minutes for a test it seems we no longer have strong confidence in. Do we still want them to be part of our test suite if they are this unreliable, they eat up significant runtime and nobody knows what's wrong?

I never managed to reproduce this but inspecting the test setup, they spawn 8 and 16 processes and GH Actions runners have 2 CPUs each (3 on OSX). I don't know how the `proc.start()` actually behaves. Is it blocking until the process is up and running or does it release during the process is still in a startup phase? Is this behaviour different on different platforms? If the process startup time is a problem here, we could try a barrier to see if this helps. The failing tests often show n_purged to be zero and this is the only thing I can think of right now which would explain this

```diff
diff --git a/distributed/tests/test_diskutils.py b/distributed/tests/test_diskutils.py
index 077a37d4..cb9d26b5 100644
--- a/distributed/tests/test_diskutils.py
+++ b/distributed/tests/test_diskutils.py
@@ -190,7 +190,8 @@ def test_locking_disabled(tmpdir):
         lock_file.assert_not_called()


-def _workspace_concurrency(base_dir, purged_q, err_q, stop_evt):
+def _workspace_concurrency(base_dir, purged_q, err_q, stop_evt, barrier):
+    barrier.wait()
     ws = WorkSpace(base_dir)
     n_purged = 0
     with captured_logger("distributed.diskutils", "ERROR") as sio:
@@ -229,15 +230,17 @@ def _test_workspace_concurrency(tmpdir, timeout, max_procs):

     # Run a bunch of child processes that will try to purge concurrently
     NPROCS = 2 if sys.platform == "win32" else max_procs
+    barrier = mp_context.Barrier(parties=NPROCS + 1)
     processes = [
         mp_context.Process(
-            target=_workspace_concurrency, args=(base_dir, purged_q, err_q, stop_evt)
+            target=_workspace_concurrency,
+            args=(base_dir, purged_q, err_q, stop_evt, barrier),
         )
         for i in range(NPROCS)
     ]
     for p in processes:
         p.start()
-
+    barrier.wait()
     n_created = 0
     n_purged = 0
     try:
```

Comment (crusaderky) [reply distributed/tests/test_diskutils.py]: thanks, I'm giving it a try

**Comment by fjetter** [comment]
@crusaderky ping once you're done. There are still failing tests but this looks already good.

File: distributed/comm/tests/test_ucx.py [hunk 5]
```diff
@@ -276,30 +276,22 @@ async def test_ucx_localcluster(processes, cleanup):
 
 
 @pytest.mark.slow
-@pytest.mark.asyncio
-async def test_stress(cleanup):
+@gen_cluster(client=True, scheduler_kwargs={"protocol": "ucx"}, timeout=240)
```
Comment (jrbourbeau) [inline distributed/comm/tests/test_ucx.py:279]: It looks like we've dropped `host=HOST` which we were previously using in the `LocalCluster` setup

Comment (crusaderky) [reply distributed/comm/tests/test_ucx.py]: reverted

File: distributed/tests/test_diskutils.py (hunk 4 above)
Comment (crusaderky) [reply distributed/tests/test_diskutils.py]: it never failed  so far! fingers crossed...

**Comment by crusaderky** [comment]
I can't see any of the changed tests failing anymore. This is ready for merge.

**Review by fjetter — APPROVED** [review]
Looks great! Thanks for taking the time to look into this

**Pull request merged**